pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,280 @@
+	default:
+		required = true
+	case "n", "no", "false", "off":
+		required = false
+	}
+	return required
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, metadata)
+	glog.Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
+	return required
+}
+
+// validate validates the persistentvolumeclaim(PVC) delete request
+func (wh *Webhook) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	req := ar.Request
+	status := &v1beta1.AdmissionResponse{}
+
+	// validates only if requested operation is DELETE and request kind is PVC
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
+		status.Allowed = true
+		return status
+	}
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	// TODO* use raw object once validation webhooks support DELETE request
+	// object as non nil value https://github.com/kubernetes/kubernetes/issues/66536
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := wh.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	if !validationRequired(ignoredNamespaces, &pvc.ObjectMeta) {
+		glog.Infof("Skipping validation for %s/%s due to policy check", pvc.Namespace, pvc.Name)
+		return &v1beta1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	// get the all CStorVolumes resources in openebs namespace(default namespace
+	// for CStorvolume resource) to check the source-volume annotation to find
+	// out if there is any clone volume exists.
+	// if source-volume annotation matches with name of PVC, failed the pvc
+	// deletion operation.
+	status.Allowed = true
+	cStorVolumes, _ := wh.Clientset.OpenebsV1alpha1().CStorVolumes("openebs").List(metav1.ListOptions{})
+	for _, cstorvolume := range cStorVolumes.Items {
+		if cstorvolume.Annotations["openebs.io/source-volume"] == pvc.Spec.VolumeName {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: "the PVC has cloned volume created",
+				Message: fmt.Sprintf("pvc %q has cloned volume PV: %q", pvc.Name, cstorvolume.Name),
```

Comment:
Can we rephrase the messages? e.g. `PVC '%s' has one or more cloned volumes`

Reply:
There is no point running through the entire list of cstor volumes if there is any volume that has this annotation matching. The Result will have the last match only.

Reply:
Ahh .. missed to add `break` after the if loop. added.